### PR TITLE
We shouldn't be subclassing Struct.new - if the file is loaded twice we get a superclass mismatch error

### PR DESCRIPTION
**Pull request by tyler-ball**
This caused an incident, so this is a simple cleanup task.

\cc @randomcamel @danielsdeleo @lamont-granquist

### CODE REVIEW

**Comment by coderanger** [comment]
:+1: lolrubby

**Comment by danielsdeleo** [comment]
IMO this isn't necessary, we just need to not load identical files (see https://github.com/chef/appbundler/pull/18) and we'll be fine. The superclass mismatch thing is just one symptom (of many; for example there's hundreds of redefined constant warnings and such) of this problem. Even if we change the chef code so things "work" if we load the same files multiple times, it'll still be an unacceptably poor experience.

I think subclassing `Struct.new` is a valid thing to do and we don't need to try to avoid it just so we can load the same file twice, because we should never be loading the same file twice.

**Comment by coderanger** [comment]
+1 for fixing this upstream, but its a relatively minor bit of bloat and const redefs are warnings vs. superclass mismatch which I think is a fatal?

**Comment by danielsdeleo** [comment]
IMO it doesn't matter that this causes an exception which leads to the termination of the program, because it only does so in the presence of a ship-blocking bug. I think it's actually kind of helpful, because it ensures that a multi-load bug will crash hard and be found (rather than spewing warnings which may not be noticed if they occur in unattended operation).

**Comment by coderanger** [comment]
Fair point :)

File: lib/chef/resource/resource_notification.rb [hunk 1]
```diff
@@ -20,7 +20,15 @@
 
 class Chef
   class Resource
-    class Notification < Struct.new(:resource, :action, :notifying_resource)
+    class Notification
+
+      attr_accessor :resource, :action, :notifying_resource
+
+      def initialize(resource, action, notifying_resource)
```
Comment (lamont-granquist) [inline lib/chef/resource/resource_notification.rb]: if back-compat is important these should all default to nil.

**Comment by lamont-granquist** [comment]
Dunno that's idiomatic use of Struct.new, but at the same time this class is more than just a Struct.new and general practice is that you shouldn't take this shortcut for things that aren't solely simple containers...  I'm kinda 50/50.

**Comment by tyler-ball** [comment]
@danielsdeleo I realize this helped us catch the underlying issue but I would rather set up a specific test for that.

I think when I originally wrote the class that it was simple enough to make extending a `Struct` make sense.  But there is enough implementation in it now that I think its should just be its own class.  Thats really why I am removing the superclass.

File: lib/chef/resource/resource_notification.rb (hunk 1 above)
Comment (tyler-ball) [inline lib/chef/resource/resource_notification.rb]: I think they should be required args.  The only place we use it is https://github.com/chef/chef/blob/master/lib/chef/resource.rb#L1527-L1535.  I know this is an API change but I don't think anyone would be using this class besides us.

**Comment by randomcamel** [comment]
This seems like a reasonable enough change: it normalizes an anomalous corner of our considerable OO hierarchy. It's not the worst thing in the world to have something in `chef/chef` that explodes when used in a bad way (like loading 2 different copies of the source), but if the only way to do that is to have a `Struct.new` descendant somewhere, we could add one such that its defensive purpose is explicit.

**Comment by lamont-granquist** [comment]
:+1:

**Comment by thommay** [comment]
yeah, let's do it :+1:

**Pull request closed without merging**